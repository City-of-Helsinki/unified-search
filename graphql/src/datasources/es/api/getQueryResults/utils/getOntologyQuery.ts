import { GraphQlToElasticLanguageMap } from '../../../../../constants';
import { type ElasticLanguage } from '../../../../../types';
import { ES_DEFAULT_INDEX } from '../../../constants';
import type { GetQueryResultsProps } from '../types';
import { getOntologyFields } from './getOntologyFields';

export const getOntologyMatchers = ({
  index = ES_DEFAULT_INDEX,
  languages = Object.values(GraphQlToElasticLanguageMap),
  query,
}: Pick<GetQueryResultsProps, 'index' | 'languages'> & { query: string }) =>
  languages.reduce<{
    [key in ElasticLanguage]?: {
      multi_match: {
        query: string;
        fields: ReturnType<typeof getOntologyFields>;
      };
    };
  }>(
    (acc, language) => ({
      ...acc,
      [language]: {
        multi_match: {
          query,
          fields: getOntologyFields(language, index),
        },
      },
    }),
    {}
  );

/** @deprecated Deprecated as unused. */
export function getOntologyQuery({
  index = ES_DEFAULT_INDEX,
  languages = Object.values(GraphQlToElasticLanguageMap),
  ontology,
}: Pick<GetQueryResultsProps, 'index' | 'languages' | 'ontology'>) {
  const ontologyMatchers = getOntologyMatchers({
    index,
    languages,
    query: ontology,
  });

  return {
    query: {
      bool: {
        should: [
          ...Object.values(
            languages.map((language) => ({
              bool: { must: [ontologyMatchers[language]] },
            }))
          ).flat(),
        ],
      },
    },
  };
}
