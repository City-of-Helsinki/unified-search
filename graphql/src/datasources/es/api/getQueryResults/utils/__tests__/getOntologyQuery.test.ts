import { describe, expect, it } from 'vitest';

import type { ElasticLanguage } from '../../../../../../types.js';
import { getOntologyMatchers, getOntologyQuery } from '../getOntologyQuery.js';

describe('getOntologyQuery', () => {
  it.each([
    { languages: [] as const satisfies ElasticLanguage[] },
    { ontology: 'test-ontology123' as const },
    {
      index: 'location' as const,
      languages: ['fi' as const, 'sv' as const],
      ontology: 'vesijumppa',
    },
  ])(
    'test output using given parameters %j',
    (parameters: Parameters<typeof getOntologyQuery>[0]) => {
      const result = getOntologyQuery(parameters);
      expect(result).toMatchSnapshot();
    }
  );
});

describe('getOntologyMatchers', () => {
  it.each([
    { query: 'test-query', languages: [] as const satisfies ElasticLanguage[] },
    { query: 'yso:1234' as const },
    {
      index: 'location' as const,
      languages: ['fi' as const, 'sv' as const],
      query: 'yso:999',
    },
  ])(
    'test output using given parameters %j',
    (parameters: Parameters<typeof getOntologyMatchers>[0]) => {
      const result = getOntologyMatchers(parameters);
      expect(result).toMatchSnapshot();
    }
  );
});
