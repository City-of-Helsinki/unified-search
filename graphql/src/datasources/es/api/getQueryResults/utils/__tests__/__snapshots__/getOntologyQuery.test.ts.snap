// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getOntologyMatchers > test output using given parameters {"index":"location","languages":["fi","sv"],"query":"yso:999"} 1`] = `
{
  "fi": {
    "multi_match": {
      "fields": [
        "links.raw_data.ontologyword_ids_enriched.extra_searchwords_fi",
        "links.raw_data.ontologyword_ids_enriched.ontologyword_fi",
        "links.raw_data.ontologytree_ids_enriched.name_fi",
        "links.raw_data.ontologytree_ids_enriched.extra_searchwords_fi",
      ],
      "query": "yso:999",
    },
  },
  "sv": {
    "multi_match": {
      "fields": [
        "links.raw_data.ontologyword_ids_enriched.extra_searchwords_sv",
        "links.raw_data.ontologyword_ids_enriched.ontologyword_sv",
        "links.raw_data.ontologytree_ids_enriched.name_sv",
        "links.raw_data.ontologytree_ids_enriched.extra_searchwords_sv",
      ],
      "query": "yso:999",
    },
  },
}
`;

exports[`getOntologyMatchers > test output using given parameters {"query":"test-query","languages":[]} 1`] = `{}`;

exports[`getOntologyMatchers > test output using given parameters {"query":"yso:1234"} 1`] = `
{
  "en": {
    "multi_match": {
      "fields": [
        "links.raw_data.ontologyword_ids_enriched.extra_searchwords_en",
        "links.raw_data.ontologyword_ids_enriched.ontologyword_en",
        "links.raw_data.ontologytree_ids_enriched.name_en",
        "links.raw_data.ontologytree_ids_enriched.extra_searchwords_en",
      ],
      "query": "yso:1234",
    },
  },
  "fi": {
    "multi_match": {
      "fields": [
        "links.raw_data.ontologyword_ids_enriched.extra_searchwords_fi",
        "links.raw_data.ontologyword_ids_enriched.ontologyword_fi",
        "links.raw_data.ontologytree_ids_enriched.name_fi",
        "links.raw_data.ontologytree_ids_enriched.extra_searchwords_fi",
      ],
      "query": "yso:1234",
    },
  },
  "sv": {
    "multi_match": {
      "fields": [
        "links.raw_data.ontologyword_ids_enriched.extra_searchwords_sv",
        "links.raw_data.ontologyword_ids_enriched.ontologyword_sv",
        "links.raw_data.ontologytree_ids_enriched.name_sv",
        "links.raw_data.ontologytree_ids_enriched.extra_searchwords_sv",
      ],
      "query": "yso:1234",
    },
  },
}
`;

exports[`getOntologyQuery > test output using given parameters {"index":"location","languages":["fi","sv"],"ontology":"vesijumppa"} 1`] = `
{
  "query": {
    "bool": {
      "should": [
        {
          "bool": {
            "must": [
              {
                "multi_match": {
                  "fields": [
                    "links.raw_data.ontologyword_ids_enriched.extra_searchwords_fi",
                    "links.raw_data.ontologyword_ids_enriched.ontologyword_fi",
                    "links.raw_data.ontologytree_ids_enriched.name_fi",
                    "links.raw_data.ontologytree_ids_enriched.extra_searchwords_fi",
                  ],
                  "query": "vesijumppa",
                },
              },
            ],
          },
        },
        {
          "bool": {
            "must": [
              {
                "multi_match": {
                  "fields": [
                    "links.raw_data.ontologyword_ids_enriched.extra_searchwords_sv",
                    "links.raw_data.ontologyword_ids_enriched.ontologyword_sv",
                    "links.raw_data.ontologytree_ids_enriched.name_sv",
                    "links.raw_data.ontologytree_ids_enriched.extra_searchwords_sv",
                  ],
                  "query": "vesijumppa",
                },
              },
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`getOntologyQuery > test output using given parameters {"languages":[]} 1`] = `
{
  "query": {
    "bool": {
      "should": [],
    },
  },
}
`;

exports[`getOntologyQuery > test output using given parameters {"ontology":"test-ontology123"} 1`] = `
{
  "query": {
    "bool": {
      "should": [
        {
          "bool": {
            "must": [
              {
                "multi_match": {
                  "fields": [
                    "links.raw_data.ontologyword_ids_enriched.extra_searchwords_fi",
                    "links.raw_data.ontologyword_ids_enriched.ontologyword_fi",
                    "links.raw_data.ontologytree_ids_enriched.name_fi",
                    "links.raw_data.ontologytree_ids_enriched.extra_searchwords_fi",
                  ],
                  "query": "test-ontology123",
                },
              },
            ],
          },
        },
        {
          "bool": {
            "must": [
              {
                "multi_match": {
                  "fields": [
                    "links.raw_data.ontologyword_ids_enriched.extra_searchwords_sv",
                    "links.raw_data.ontologyword_ids_enriched.ontologyword_sv",
                    "links.raw_data.ontologytree_ids_enriched.name_sv",
                    "links.raw_data.ontologytree_ids_enriched.extra_searchwords_sv",
                  ],
                  "query": "test-ontology123",
                },
              },
            ],
          },
        },
        {
          "bool": {
            "must": [
              {
                "multi_match": {
                  "fields": [
                    "links.raw_data.ontologyword_ids_enriched.extra_searchwords_en",
                    "links.raw_data.ontologyword_ids_enriched.ontologyword_en",
                    "links.raw_data.ontologytree_ids_enriched.name_en",
                    "links.raw_data.ontologytree_ids_enriched.extra_searchwords_en",
                  ],
                  "query": "test-ontology123",
                },
              },
            ],
          },
        },
      ],
    },
  },
}
`;
